### ISSUE TITLE
Primary key of device 'PetersonTestBed' was marked as invalid. No other device keys available

### ISSUE THREAD

**Issue by tikluganguly** [Azure/azure-iot-sdk-c#106]
- **OS and version used:** Linux TC3G-152157731004 3.10.94-rt102/STW-v1_03r0+ #1 Thu Aug 4 10:05:38 CEST 2016 ppc GNU/Linux

- **SDK version used:** IoT Hub SDK for C, version 1.1.10

# Description of the issue:
I have taken code from the following sample
https://github.com/Azure/azure-iot-sdk-c/blob/master/iothub_client/samples/iothub_client_sample_amqp/iothub_client_sample_amqp.c

The code is almost identical including the messages. The only thing it lacks is it does not have a receive message callback set. 

While running the code I am getting the following error.

# Console log of the issue:
```
Starting the IoTHub client call...
Info: IoT Hub SDK for C, version 1.1.10
IoTHubClient_SendEventAsync accepted data for transmission to IoT Hub.
IoTHubClient_SendEventAsync accepted data for transmission to IoT Hub.
IoTHubClient_SendEventAsync accepted data for transmission to IoT Hub.
IoTHubClient_SendEventAsync accepted data for transmission to IoT Hub.
IoTHubClient_SendEventAsync accepted data for transmission to IoT Hub.
Error: Time:Mon Apr  3 23:44:30 2017 File:/home/tiklu/azure/azure-iot-sdk-c/iothub_client/src/iothubtransport_amqp_cbs_auth.c Func:mark_current_device_key_as_invalid Line:216 Primary key of device 'PetersonTestBed' was marked as invalid. No other device keys available
Error: Time:Mon Apr  3 23:44:30 2017 File:/home/tiklu/azure/azure-iot-sdk-c/iothub_client/src/iothubtransport_amqp_common.c Func:IoTHubTransport_AMQP_Common_Device_DoWork Line:979 Failed performing DoWork for device 'PetersonTestBed' (device reported state 5; number of previous failures: 0)
Error: Time:Mon Apr  3 23:44:30 2017 File:/home/tiklu/azure/azure-iot-sdk-c/iothub_client/src/iothubtransport_amqp_common.c Func:IoTHubTransport_AMQP_Common_Device_DoWork Line:979 Failed performing DoWork for device 'PetersonTestBed' (device reported state 4; number of previous failures: 1)
Error: Time:Mon Apr  3 23:44:30 2017 File:/home/tiklu/azure/azure-iot-sdk-c/iothub_client/src/iothubtransport_amqp_common.c Func:IoTHubTransport_AMQP_Common_Device_DoWork Line:979 Failed performing DoWork for device 'PetersonTestBed' (device reported state 4; number of previous failures: 2)
Error: Time:Mon Apr  3 23:44:30 2017 File:/home/tiklu/azure/azure-iot-sdk-c/iothub_client/src/iothubtransport_amqp_common.c Func:IoTHubTransport_AMQP_Common_Device_DoWork Line:979 Failed performing DoWork for device 'PetersonTestBed' (device reported state 4; number of previous failures: 3)
Error: Time:Mon Apr  3 23:44:30 2017 File:/home/tiklu/azure/azure-iot-sdk-c/iothub_client/src/iothubtransport_amqp_common.c Func:IoTHubTransport_AMQP_Common_Device_DoWork Line:979 Failed performing DoWork for device 'PetersonTestBed' (device reported state 4; number of previous failures: 4)
Error: Time:Mon Apr  3 23:44:30 2017 File:/home/tiklu/azure/azure-iot-sdk-c/iothub_client/src/iothubtransport_amqp_common.c Func:IoTHubTransport_AMQP_Common_DoWork Line:1343 Device 'PetersonTestBed' reported a critical failure; connection retry will be triggered.
Error: Time:Mon Apr  3 23:44:30 2017 File:/home/tiklu/azure/azure-iot-sdk-c/iothub_client/src/iothubtransport_amqp_common.c Func:IoTHubTransport_AMQP_Common_DoWork Line:1304 An error occured on AMQP connection. The connection will be restablished.
Error: Time:Mon Apr  3 23:44:30 2017 File:/home/tiklu/azure/azure-iot-sdk-c/iothub_client/src/iothubtransport_amqp_connection.c Func:on_cbs_open_complete Line:149 CBS open failed
Error: Time:Mon Apr  3 23:44:30 2017 File:/home/tiklu/azure/azure-iot-sdk-c/iothub_client/src/iothubtransport_amqp_cbs_auth.c Func:on_cbs_put_token_complete_callback Line:272 CBS reported status code 0, error: '(null)' for put-token operation for device 'PetersonTestBed'
-> [CLOSE]* {}
```

the connection string works fine with a python client.

**Comment by tameraw** [comment]
@tikluganguly - Can you please confirm that the connection string copied from python to the C client is indeed a valid one? 
If you create a new device and use its connection string in the C client code you have, does it work?

**Comment by tikluganguly** [comment]
@tameraw - Yes I have tested the code with two connection strings from two different devices of the same IoTHub. I have copied the connection string directly from azure IoT device explorer. The devices are enabled and the connection string has been used previously to send data from the python code.

**Comment by tikluganguly** [comment]
I just copied the code from https://github.com/Azure/azure-iot-sdk-c/blob/master/iothub_client/samples/iothub_client_sample_amqp/iothub_client_sample_amqp.c and added my connection string. This code also gives a similar output. 

# ./azure1
Starting the IoTHub client sample AMQP...
Info: IoT Hub SDK for C, version 1.1.10
IoTHubClient_SetMessageCallback...successful.
IoTHubClient_SendEventAsync accepted data for transmission to IoT Hub.
IoTHubClient_SendEventAsync accepted data for transmission to IoT Hub.
IoTHubClient_SendEventAsync accepted data for transmission to IoT Hub.
IoTHubClient_SendEventAsync accepted data for transmission to IoT Hub.
IoTHubClient_SendEventAsync accepted data for transmission to IoT Hub.
Error: Time:Wed Apr  5 15:43:46 2017 File:/home/tiklu/azure/azure-iot-sdk-c/iothub_client/src/iothubtransport_amqp_cbs_auth.c Func:mark_current_device_key_as_invalid Line:216 Primary key of device 'PetersonTestBed' was marked as invalid. No other device keys available
Error: Time:Wed Apr  5 15:43:46 2017 File:/home/tiklu/azure/azure-iot-sdk-c/iothub_client/src/iothubtransport_amqp_common.c Func:IoTHubTransport_AMQP_Common_Device_DoWork Line:979 Failed performing DoWork for device 'PetersonTestBed' (device reported state 5; number of previous failures: 0)
Error: Time:Wed Apr  5 15:43:46 2017 File:/home/tiklu/azure/azure-iot-sdk-c/iothub_client/src/iothubtransport_amqp_common.c Func:IoTHubTransport_AMQP_Common_Device_DoWork Line:979 Failed performing DoWork for device 'PetersonTestBed' (device reported state 4; number of previous failures: 1)
Error: Time:Wed Apr  5 15:43:46 2017 File:/home/tiklu/azure/azure-iot-sdk-c/iothub_client/src/iothubtransport_amqp_common.c Func:IoTHubTransport_AMQP_Common_Device_DoWork Line:979 Failed performing DoWork for device 'PetersonTestBed' (device reported state 4; number of previous failures: 2)
Error: Time:Wed Apr  5 15:43:46 2017 File:/home/tiklu/azure/azure-iot-sdk-c/iothub_client/src/iothubtransport_amqp_common.c Func:IoTHubTransport_AMQP_Common_Device_DoWork Line:979 Failed performing DoWork for device 'PetersonTestBed' (device reported state 4; number of previous failures: 3)
Error: Time:Wed Apr  5 15:43:46 2017 File:/home/tiklu/azure/azure-iot-sdk-c/iothub_client/src/iothubtransport_amqp_common.c Func:IoTHubTransport_AMQP_Common_Device_DoWork Line:979 Failed performing DoWork for device 'PetersonTestBed' (device reported state 4; number of previous failures: 4)
Error: Time:Wed Apr  5 15:43:46 2017 File:/home/tiklu/azure/azure-iot-sdk-c/iothub_client/src/iothubtransport_amqp_common.c Func:IoTHubTransport_AMQP_Common_DoWork Line:1343 Device 'PetersonTestBed' reported a critical failure; connection retry will be triggered.
Error: Time:Wed Apr  5 15:43:46 2017 File:/home/tiklu/azure/azure-iot-sdk-c/iothub_client/src/iothubtransport_amqp_common.c Func:IoTHubTransport_AMQP_Common_DoWork Line:1304 An error occured on AMQP connection. The connection will be restablished.
Error: Time:Wed Apr  5 15:43:46 2017 File:/home/tiklu/azure/azure-iot-sdk-c/iothub_client/src/iothubtransport_amqp_connection.c Func:on_cbs_open_complete Line:149 CBS open failed
Error: Time:Wed Apr  5 15:43:46 2017 File:/home/tiklu/azure/azure-iot-sdk-c/iothub_client/src/iothubtransport_amqp_cbs_auth.c Func:on_cbs_put_token_complete_callback Line:272 CBS reported status code 0, error: '(null)' for put-token operation for device 'PetersonTestBed'
-> [CLOSE]* {}
Error: Time:Wed Apr  5 15:43:46 2017 File:/home/tiklu/azure/azure-iot-sdk-c/iothub_client/src/iothubtransport_amqp_common.c Func:IoTHubTransport_AMQP_Common_Device_DoWork Line:979 Failed performing DoWork for device 'PetersonTestBed' (device reported state 4; number of previous failures: 0)
Error: Time:Wed Apr  5 15:43:46 2017 File:/home/tiklu/azure/azure-iot-sdk-c/iothub_client/src/iothubtransport_amqp_common.c Func:IoTHubTransport_AMQP_Common_Device_DoWork Line:979 Failed performing DoWork for device 'PetersonTestBed' (device reported state 4; number of previous failures: 1)
Error: Time:Wed Apr  5 15:43:46 2017 File:/home/tiklu/azure/azure-iot-sdk-c/iothub_client/src/iothubtransport_amqp_common.c Func:IoTHubTransport_AMQP_Common_Device_DoWork Line:979 Failed performing DoWork for device 'PetersonTestBed' (device reported state 4; number of previous failures: 2)
Error: Time:Wed Apr  5 15:43:46 2017 File:/home/tiklu/azure/azure-iot-sdk-c/iothub_client/src/iothubtransport_amqp_common.c Func:IoTHubTransport_AMQP_Common_Device_DoWork Line:979 Failed performing DoWork for device 'PetersonTestBed' (device reported state 4; number of previous failures: 3)
Error: Time:Wed Apr  5 15:43:46 2017 File:/home/tiklu/azure/azure-iot-sdk-c/iothub_client/src/iothubtransport_amqp_common.c Func:IoTHubTransport_AMQP_Common_Device_DoWork Line:979 Failed performing DoWork for device 'PetersonTestBed' (device reported state 4; number of previous failures: 4)
Error: Time:Wed Apr  5 15:43:46 2017 File:/home/tiklu/azure/azure-iot-sdk-c/iothub_client/src/iothubtransport_amqp_common.c Func:IoTHubTransport_AMQP_Common_DoWork Line:1343 Device 'PetersonTestBed' reported a critical failure; connection retry will be triggered.
Error: Time:Wed Apr  5 15:43:46 2017 File:/home/tiklu/azure/azure-iot-sdk-c/iothub_client/src/iothubtransport_amqp_common.c Func:IoTHubTransport_AMQP_Common_DoWork Line:1304 An error occured on AMQP connection. The connection will be restablished.
Error: Time:Wed Apr  5 15:43:46 2017 File:/home/tiklu/azure/azure-iot-sdk-c/iothub_client/src/iothubtransport_amqp_connection.c Func:on_cbs_open_complete Line:149 CBS open failed
-> [CLOSE]* {}

**Comment by tikluganguly** [comment]
I am not sure what other info I should provide. So here are some more info about it. The c SDK is cross-compiled to run on a ppc based device. I have used the following commands to compile it


powerpc-stw-linux-uclibc-gcc -std=c99 -I/opt/powerpc-gcc-4.4.3-uclib/powerpc-stw-linux-uclibc/sysroot/usr/include -I/opt/powerpc-gcc-4.4.3-uclib/powerpc-stw-linux-uclibc/sysroot/include -I/home/tiklu/azure/azure-iot-sdk-c/c-utility/inc -I/home/tiklu/azure/azure-iot-sdk-c/iothub_client/inc  -c -o azure1.o azure1.c

powerpc-stw-linux-uclibc-gcc -Wall -std=c99 -L/opt/powerpc-gcc-4.4.3-uclib/powerpc-stw-linux-uclibc/sysroot/usr/lib/ -L/opt/powerpc-gcc-4.4.3-uclib/powerpc-stw-linux-uclibc/sysroot/lib -luClibc-0.9.33.2  -lpthread -lssl -lcrypto -lcurl -luuid -lm -lwebsockets azure1.o /home/tiklu/azure/azure-iot-sdk-c/cmake/iotsdk_linux/iothub_client/libiothub_client_amqp_transport.a /home/tiklu/azure/azure-iot-sdk-c/cmake/iotsdk_linux/iothub_client/libiothub_client.a /home/tiklu/azure/azure-iot-sdk-c/cmake/iotsdk_linux/serializer/libserializer.a /home/tiklu/azure/azure-iot-sdk-c/cmake/iotsdk_linux/uamqp/libuamqp.a /home/tiklu/azure/azure-iot-sdk-c/cmake/iotsdk_linux/c-utility/libaziotsharedutil.a -oazure1

The code in azure1.c is pasted from https://github.com/Azure/azure-iot-sdk-c/blob/master/iothub_client/samples/iothub_client_sample_amqp/iothub_client_sample_amqp.c

**Comment by tameraw** [comment]
@tikluganguly - Thanks for the additional info. This needs further investigations as we don't have this toolchain in our test environment.

**Comment by markrad** [comment]
HI @tikluganguly 

Can you tell me which datacenter is hosting your IoT hub please?

_Mark Radbourne MSFT_

**Comment by tikluganguly** [comment]
Hi @markrad,

        sorry for the late reply. I have the iothub at "West US". 

Thanks
Tiklu

**Comment by andygikling** [comment]
I'm having the exact same issue as tikluganguly.

I put my connection string into the example "iothub_client_sample_amqp.c."  I can test the code on Ubuntu 16.04 and it runs correctly - I can see the data ingress into my Azure IoTHub using the web portal.

So I know my code works and my connection string is valid.

Then I cross compile for my target which is a Variscite Dart SOM (Yocto Linux 4.1 built using the Yocto supplied ARM cross compiler "arm-poky-linux-gnueabi") and I get the error message from above "Primary key of device 'MyIoTDeviceNameHere' was marked as invalid. No other device keys available"

I'm wondering if this has anything to do with my target not having x509 certificates setup correctly (I don't believe my target doesn't have x509).  Is this required?

Thanks,

~Andy

**Comment by ewertons** [comment]
Hi @markrad , just showing up to share a couple of thoughts.

For the x509 question above, no they are not needed for authentication when you are using SAS tokens (i.e., passing a connection string with primary key information).

Make sure you are calling SetOption("TrustedCerts"... if running on a platform other than Windows.

If this does not fix it, I would use a debugger to break into the code when the SASToken_Create is invoked and check what is the resulting string (on both the C and Python samples, to compare).

**Comment by andygikling** [comment]
@ewertons,

Thanks for the tip that worked!  Using #define MBED_BUILD_TIMESTAMP in my sample, brings in certs.h which is key.  My AMQP sample runs on the iMX6 SOM now.

Looking through the commit history I see that a certs.h file was updated by Andrei Porumb under /certs/cert.h on 3/8/2017 - I'm not sure if this is the right one to use, but it fixed my problem.  

What's interesting is I would expect the MQTT sample to also work on my SOM but it does not.

I'm using origin/master as of 03/22/2017 (aa9a52fb).
Here is the error I get (showing 1 retry):

IoTHubClient_SendEventAsync accepted data for transmission to IoT Hub.
-> 19:36:21 CONNECT | VER: 4 | KEEPALIVE: 240 | FLAGS: 192 | USERNAME: iotHubSim.azure-devices.net/customerXdevice0001/api-version=2016-11-14&DeviceClientType=iothubclient%2F1.1.9 | PWD: XXXX | CLEAN: 0
<- 19:36:22 CONNACK | SESSION_PRESENT: false | RETURN_CODE: 0x5
Error: Time:Tue Apr 11 19:36:22 2017 File:/opt/iotHubSim/Git/iotHubSim.Client/Lib/azure-iot-sdk-c/iothub_client/src/iothubtransport_mqtt_common.c Func:mqtt_operation_complete_callback Line:1411 Connection Not Accepted: 0x5: Not Authorized
-> 19:36:22 DISCONNECT
Error: Time:Tue Apr 11 19:36:22 2017 File:/opt/iotHubSim/Git/iotHubSim.Client/Lib/azure-iot-sdk-c/c-utility/src/tlsio_openssl.c Func:decode_ssl_received_bytes Line:648 SSL channel closed in decode_ssl_received_bytes.
Error: Time:Tue Apr 11 19:36:22 2017 File:/opt/iotHubSim/Git/iotHubSim.Client/Lib/azure-iot-sdk-c/c-utility/src/tlsio_openssl.c Func:on_underlying_io_bytes_received Line:705 Error in decode_ssl_received_bytes.
Info: Evaluated delay 0 at 1 attempt to retry

-> 19:36:24 CONNECT | VER: 4 | KEEPALIVE: 240 | FLAGS: 192 | USERNAME: iotHubSim.azure-devices.net/customerXdevice0001/api-version=2016-11-14&DeviceClientType=iothubclient%2F1.1.9 | PWD: XXXX | CLEAN: 0
<- 19:36:25 CONNACK | SESSION_PRESENT: false | RETURN_CODE: 0x5
Error: Time:Tue Apr 11 19:36:25 2017 File:/opt/iotHubSim/Git/iotHubSim.Client/Lib/azure-iot-sdk-c/iothub_client/src/iothubtransport_mqtt_common.c Func:mqtt_operation_complete_callback Line:1411 Connection Not Accepted: 0x5: Not Authorized
-> 19:36:25 DISCONNECT
Error: Time:Tue Apr 11 19:36:25 2017 File:/opt/iotHubSim/Git/iotHubSim.Client/Lib/azure-iot-sdk-c/c-utility/src/tlsio_openssl.c Func:decode_ssl_received_bytes Line:648 SSL channel closed in decode_ssl_received_bytes.
Error: Time:Tue Apr 11 19:36:25 2017 File:/opt/iotHubSim/Git/iotHubSim.Client/Lib/azure-iot-sdk-c/c-utility/src/tlsio_openssl.c Func:on_underlying_io_bytes_received Line:705 Error in decode_ssl_received_bytes.
Info: Evaluated delay 0 at 2 attempt to retry

Is MQTT also dependent on SAS tokens?  Do I need a different certificate?  Is my server not listed in the certs.h?

Also, why do you use the define "MBED_BUILD_TIMESTAMP?" - if every embedded Linux device is going to need these static certs then why wouldn't that be called something like "USE_STATIC_CERTS?"  Also a CMAKE switch for this would be nice.

Thanks,

~Andy

**Comment by andygikling** [comment]
Bump,

I'm sorry I was wrong in my last post.  MQTT is working on my SOM.  Using SetOption() "TrustedCerts" makes it work.

Thanks,

~Andy

**Comment by tikluganguly** [comment]
great to know that @andygikling problem got solved. any update for me guys :)

**Comment by markrad** [comment]
Hi @andygikling,

The MBED define is a historical artifact. When the C SDK was first written the only platform that did not support a certificate store was the MBED. Unfortunately, as you are now aware, many other platforms also do not have a certificate store so the name of the define is now somewhat unintuitive. I agree that a CMake switch would be a better way to address this. I'll raise the question with the product group.

_Mark Radbourne MSFT_

**Comment by markrad** [comment]
Hi @tikluganguly,

Did you try passing the certificates to the SDK in the manner that @andygikling did? If you don't have a certificate root store on your device then the connection will often fail at the TLS negotiation. Another issue that may cause this is if the current time on your device is inaccurate. This can cause improperly constructed SAS tokens and issues regarding the certificate's period of validity. You should also make sure that you are not being blocked by a firewall. AMQP uses port 5671, MQTT uses 8882 and HTTPS uses the normal port 443. If you are not sure then, if you have a telnet client on you device, you can test it by entering 
```
telnet <Your Hub Name>.azure-devices.com 5671
```
It should be able to connect and will timeout after a short while.

_Mark Radbourne MSFT_

**Comment by tikluganguly** [comment]
Hi @markrad,

                I did not try that. It does seem like a plausible issue. Let me try using the cert and I will get back to you on this. 

Regards
Tiklu

**Comment by tikluganguly** [comment]
Hi @markrad ,

          I am getting the following message

# telnet <myhubname>.azure-devices.net 5671
Connection closed by foreign host

It does not say anything about connection timeout. So does this means I am missing the certificate?

Regards
Tiklu

**Comment by tikluganguly** [comment]
Hi Guys,

         I did notice that the gateway had a wrong date. so I fixed it. But I am still clueless about how to set the certificate. Can you please help me.

Regards
Tiklu

**Comment by markrad** [comment]
Hi @tikluganguly,

Seeing the connection closed is normal. It does suggest though that your device was able to communicate successfully with the IoT hub.

Can you tell me about the device you are using? Do you have OpenSSL on this device? Small footprint Linux versions such as OpenWrt for example do not have a root certificate store so it is necessary to pass them to the SDK. You can tell if you have a root certificate store, for OpenSSL at least, by entering
```
openssl version -d
```
If the above command shows you a directory then try the command
```
openssl s_client -connect <your hub name>.azure-devices.net:5671
```
If the SSL negotiation is successful you will see the line Verify return code: 0 (ok). at the bottom of the output. You can press Ctrl-C to terminate the program. If you see anything else please post it here.

_Mark Radbourne MSFT_

**Comment by andygikling** [comment]
@tikluganguly 

If your device doesn't have a root certificate store (it might not!) then you need to compile the certificates into your sample code.

On Azure AMQP samples in this repo, have a look for the ifdef "MBED_BUILD_TIMESTAMP" and you will see if that gets defined the sample does an #include "certs.h"

That file contains your static certificates.  It can be found in this repo at "/certs/certs.h"

**Comment by tikluganguly** [comment]
Hi @markrad,
Thanks a lot for your guidance. It does seem the device does have OpenSSL because the output of the first command is
   
          OPENSSLDIR: "/etc/ssl"

But the second command ends in 

       Verify return code: 20 (unable to get local issuer certificate)

So I am assuming I may have to either create a local certificate or I need to pass a certificate as per the instruction is given by @andygikling. 

Thanks
Tiklu

**Comment by tikluganguly** [comment]
Dear @andygikling,

Thanks for the guidance. Let me give this a try. 

Thanks
Tiklu

**Comment by tikluganguly** [comment]
Hi @markrad,
                I can see a ca-certificates.crt inside /etc/ssl/certs and following is the output of the command with session key and other values retracted.

    openssl  s_client -CApath /etc/ssl/certs -connect mazikdev2.azure-devices.net:5671

CONNECTED(00000003)
depth=1 C = US, ST = Washington, L = Redmond, O = Microsoft Corporation, OU = Microsoft IT, CN = Microsoft IT SSL SHA2
verify error:num=20:unable to get local issuer certificate
verify return:0

Certificate chain
 0 s:/CN=*.azure-devices.net
   i:/C=US/ST=Washington/L=Redmond/O=Microsoft Corporation/OU=Microsoft IT/CN=Microsoft IT SSL SHA2
 1 s:/C=US/ST=Washington/L=Redmond/O=Microsoft Corporation/OU=Microsoft IT/CN=Microsoft IT SSL SHA2
   i:/C=IE/O=Baltimore/OU=CyberTrust/CN=Baltimore CyberTrust Root

Server certificate
BEGIN CERTIFICATE-----
<certificate>
END CERTIFICATE-----
subject=/CN=*.azure-devices.net
issuer=/C=US/ST=Washington/L=Redmond/O=Microsoft Corporation/OU=Microsoft IT/CN=Microsoft IT SSL SHA2

No client certificate CA names sent

SSL handshake has read 3677 bytes and written 546 bytes

New, TLSv1/SSLv3, Cipher is ECDHE-RSA-AES256-SHA384
Server public key is 2048 bit
Secure Renegotiation IS supported
Compression: NONE
Expansion: NONE
SSL-Session:
    Protocol  : TLSv1.2
    Cipher    : ECDHE-RSA-AES256-SHA384
    Session-ID: <alphanum value> 
    Session-ID-ctx:
    Master-Key: <alphanum value>
    Key-Arg   : None
    PSK identity: None
    PSK identity hint: None
    SRP username: None
    Start Time: 1492438880
    Timeout   : 300 (sec)
    Verify return code: 20 (unable to get local issuer certificate)

**Comment by tikluganguly** [comment]
Hi Guys,

Finally I was able to receive message using the explicite certificate. But I would definitely like to know how to fix my ssl. 

Below is the output

Starting the IoTHub client sample AMQP...
Info: IoT Hub SDK for C, version 1.1.10
IoTHubClient_SetMessageCallback...successful.
IoTHubClient_SendEventAsync accepted data for transmission to IoT Hub.
IoTHubClient_SendEventAsync accepted data for transmission to IoT Hub.
IoTHubClient_SendEventAsync accepted data for transmission to IoT Hub.
IoTHubClient_SendEventAsync accepted data for transmission to IoT Hub.
IoTHubClient_SendEventAsync accepted data for transmission to IoT Hub.
-> Header (AMQP 0.1.0.0)
<- Header (AMQP 0.1.0.0)
-> [OPEN]* {3d7e8a68-ad44-43a3-aa26-e61a480f88f7,MazikDev2.azure-devices.net,4294967295,65535,240000}
<- [OPEN]* {DeviceGateway_254351d564664ac7947bca16d51a70b7,10.0.0.31,65536,8191,240000,NULL,NULL,NULL,NULL,NULL}
-> [BEGIN]* {NULL,0,4294967295,100,4294967295}
<- [BEGIN]* {0,1,100,5000,262143,NULL,NULL,NULL}
-> [ATTACH]* {cbs-sender,0,false,0,0,* {$cbs},* {$cbs},NULL,NULL,0,65535}
-> [ATTACH]* {cbs-receiver,1,true,0,0,* {$cbs},* {$cbs},NULL,NULL,NULL,65535}
<- [ATTACH]* {cbs-sender,0,true,0,0,* {$cbs,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL},* {$cbs,NULL,NULL,NULL,NULL,NULL,NULL},NULL,NULL,NULL,65535,NULL,NULL,NULL}
<- [FLOW]* {0,100,1,5000,0,0,100,0,NULL,false,NULL}
<- [ATTACH]* {cbs-receiver,1,false,0,0,* {$cbs,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL},* {$cbs,NULL,NULL,NULL,NULL,NULL,NULL},NULL,NULL,0,65535,NULL,NULL,NULL}
-> [FLOW]* {1,4294967295,0,100,1,0,10000}
-> [TRANSFER]* {0,0,<00 00 00 01>,0,false,false}
<- [DISPOSITION]* {true,0,NULL,true,* {},NULL}
<- [TRANSFER]* {1,0,<01 00 00 00>,0,NULL,false,NULL,NULL,NULL,NULL,false}
-> [DISPOSITION]* {true,0,0,true,* {}}
-> [ATTACH]* {link-snd-myFirstDevice-0b4cf767-1af6-4a9b-a834-02164a211498,2,false,0,0,* {link-snd-myFirstDevice-0b4cf767-1af6-4a9b-a834-02164a211498-source},* {amqps://MazikDev2.azure-devices.net/devices/myFirstDevice/messages/events},NULL,NULL,0,18446744073709551615,NULL,NULL,{[com.microsoft:client-version:iothubclient/1.1.10]}}
<- [ATTACH]* {link-snd-myFirstDevice-0b4cf767-1af6-4a9b-a834-02164a211498,2,true,0,NULL,* {link-snd-myFirstDevice-0b4cf767-1af6-4a9b-a834-02164a211498-source,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL},* {amqps://MazikDev2.azure-devices.net/devices/myFirstDevice/messages/events,NULL,NULL,NULL,NULL,NULL,NULL},NULL,NULL,NULL,1048576,NULL,NULL,{[com.microsoft:client-version:iothubclient/1.1.10]}}
-> [ATTACH]* {link-rcv-myFirstDevice-03b99846-8f7d-4012-ac14-3a6d4ffdc3b0,3,true,0,0,* {amqps://MazikDev2.azure-devices.net/devices/myFirstDevice/messages/devicebound},* {link-rcv-myFirstDevice-03b99846-8f7d-4012-ac14-3a6d4ffdc3b0-target},NULL,NULL,NULL,65536,NULL,NULL,{[com.microsoft:client-version:iothubclient/1.1.10]}}
<- [FLOW]* {1,100,2,4294967295,2,0,1000,0,NULL,false,NULL}
-> [TRANSFER]* {2,1,<00 00 00 01>,0,false,false}
-> [TRANSFER]* {2,2,<00 00 00 02>,0,false,false}
-> [TRANSFER]* {2,3,<00 00 00 03>,0,false,false}
-> [TRANSFER]* {2,4,<00 00 00 04>,0,false,false}
-> [TRANSFER]* {2,5,<00 00 00 05>,0,false,false}
<- [ATTACH]* {link-rcv-myFirstDevice-03b99846-8f7d-4012-ac14-3a6d4ffdc3b0,3,false,NULL,1,* {amqps://MazikDev2.azure-devices.net/devices/myFirstDevice/messages/devicebound,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL},* {link-rcv-myFirstDevice-03b99846-8f7d-4012-ac14-3a6d4ffdc3b0-target,NULL,NULL,NULL,NULL,NULL,NULL},NULL,NULL,0,65536,NULL,NULL,{[com.microsoft:client-version:iothubclient/1.1.10]}}
-> [FLOW]* {2,4294967294,6,94,3,0,10000}
<- [DISPOSITION]* {true,1,2,true,* {},NULL}
Confirmation[0] received for message tracking id = 0 with result = IOTHUB_CLIENT_CONFIRMATION_OK
Confirmation[1] received for message tracking id = 1 with result = IOTHUB_CLIENT_CONFIRMATION_OK
<- [DISPOSITION]* {true,3,NULL,true,* {},NULL}
Confirmation[2] received for message tracking id = 2 with result = IOTHUB_CLIENT_CONFIRMATION_OK
<- [DISPOSITION]* {true,5,NULL,true,* {},NULL}
Confirmation[3] received for message tracking id = 4 with result = IOTHUB_CLIENT_CONFIRMATION_OK
<- [DISPOSITION]* {true,4,NULL,true,* {},NULL}
Confirmation[4] received for message tracking id = 3 with result = IOTHUB_CLIENT_CONFIRMATION_OK

Regards
Tiklu

**Comment by markrad** [comment]
Hi @tikluganguly,

Since you have a root certificate store then it is likely that it does not contain the root certificate required to verify the certificate returned by the Azure IoT Hub. You can find out if you have this certificate by looking in the certs directory which, in your case, should be /etc/ssl/certs. Search for a certificate with a name that begins with "Baltimore". Here is an example from my Ubuntu machine:
```
markrad@MARKRADUBU:~$ ls -al /etc/ssl/certs/ | grep -i baltimore
lrwxrwxrwx 1 root root     29 Mar 14 16:56 3ad48a91.0 -> Baltimore_CyberTrust_Root.pem
lrwxrwxrwx 1 root root     29 Mar 14 16:56 653b494a.0 -> Baltimore_CyberTrust_Root.pem
lrwxrwxrwx 1 root root     64 Mar  8 14:42 Baltimore_CyberTrust_Root.pem -> /usr/share/ca-certificates/mozilla/Baltimore_CyberTrust_Root.crt
```
Piping the output to grep will find that a symbolic link to the actual certificate and also two other symbolic links with hex names that are used by OpenSSL to find the certificate it needs via a hashing mechanism. It seems likely that this certificate is missing on your system. 

Unfortunately adding a new root certificate is not just as simple as copying it to your root certificate directory. You will also need to generate the hash file symbolic links. Take a look at this page to get started: http://www.gagravarr.org/writing/openssl-certs/others.shtml. 

I will warn you though that I have come across some embedded Linux systems that refused to use the root certificate store no matter what I did.

_Mark Radbourne MSFT_

**Comment by tikluganguly** [comment]
Dear @markrad,
              Thanks for your instructions. I will give it a try and see if I can make the root cert work. When I started reading your response I was thinking "easy peasy, I will just copy those certs from another machine". Then I came across your warning :D. Anyways I will give it a try. I am closing out this issue as the main problem that I was having is solved. And I don't think if the device does not support SSL certs then it has anything to do directly with the azure sdk. 

So thanks, @markrad and  @andygikling for your valuable time. 

Regards
Tiklu

**Comment by markrad** [comment]
You're welcome.

One quick short cut - the OpenSSL hash function will generate the same hash for a certificate regardless of the machine it is being run upon. You actually just need to create the two symbolic links shown above in my sample output once you have the Baltimore Cyber Trust certificate on your machine. I reviewed a couple of different machines running different versions of Linux and the symbolic link file names match.

_Mark Radbourne MSFT_

**Comment by tikluganguly** [comment]
cool, I think I will give it a try. I will copy the certs from my ubuntu box to the gateway.
